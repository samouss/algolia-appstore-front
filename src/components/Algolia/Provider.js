import { Component, Children } from 'react';
import PropTypes from 'prop-types';
import algoliasearch from 'algoliasearch';
import algoliasearchHelper from 'algoliasearch-helper';
import { highlightPreTag, highlightPostTag } from './utils';

export const ContextTypes = {
  algoliaClient: PropTypes.object.isRequired,
  algoliaHelper: PropTypes.object.isRequired,
};

class Provider extends Component {
  constructor(props) {
    super(props);

    this.client = algoliasearch(props.appId, props.apiKey);
    this.helper = algoliasearchHelper(this.client, props.indexName, props.options);
  }

  getChildContext() {
    return {
      algoliaClient: this.client,
      algoliaHelper: this.helper,
    };
  }

  componentDidMount() {
    this.helper
      .setQueryParameter('highlightPreTag', highlightPreTag)
      .setQueryParameter('highlightPostTag', highlightPostTag)
      .search();
  }

  render() {
    return Children.only(this.props.children);
  }
}

Provider.propTypes = {
  appId: PropTypes.string.isRequired,
  apiKey: PropTypes.string.isRequired,
  indexName: PropTypes.string.isRequired,
  children: PropTypes.node.isRequired,
  // eslint-disable-next-line react/forbid-prop-types
  options: PropTypes.object,
};

Provider.defaultProps = {
  options: {},
};

Provider.childContextTypes = ContextTypes;

export default Provider;
