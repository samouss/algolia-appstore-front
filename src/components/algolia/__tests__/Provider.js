import React from 'react';
import { shallow } from 'enzyme';
import algoliasearch from 'algoliasearch';
import algoliasearchHelper from 'algoliasearch-helper';
import Provider from '../Provider';

jest.mock('algoliasearch', () => jest.fn(() => ({
  name: 'algolia-client',
})));

jest.mock('algoliasearch-helper', () => jest.fn(() => ({
  name: 'algolia-helper',
})));

describe('algolia', () => {
  describe('<Provider />', () => {
    const defaultProps = {
      appId: 'ALGOLIA_APP_ID',
      apiKey: 'ALGOLIA_API_KEY',
      indexName: 'ALGOLIA_INDEX_NAME',
    };

    it('expect to render', () => {
      const props = {
        ...defaultProps,
      };

      const component = shallow(
        <Provider
          {...props}
        >
          <div>Content</div>
        </Provider>,
      );

      expect(component).toMatchSnapshot();
    });

    it('expect to instanciate client and helper', () => {
      const props = {
        ...defaultProps,
      };

      shallow(
        <Provider
          {...props}
        >
          <div>Content</div>
        </Provider>,
      );

      expect(algoliasearch).toHaveBeenCalledWith(
        props.appId,
        props.apiKey,
      );

      expect(algoliasearchHelper).toHaveBeenCalledWith(
        { name: 'algolia-client' },
        props.indexName,
      );
    });

    it('expect to bind client and helper to context', () => {
      const props = {
        ...defaultProps,
      };

      const component = shallow(
        <Provider
          {...props}
        >
          <div>Content</div>
        </Provider>,
      );

      const expectation = {
        algoliaClient: { name: 'algolia-client' },
        algoliaHelper: { name: 'algolia-helper' },
      };

      const actual = component.instance().getChildContext();

      expect(actual).toEqual(expectation);
    });

    it('expect to call search function onDidMount', () => {
      const props = {
        ...defaultProps,
      };

      algoliasearchHelper.mockImplementationOnce(() => ({
        search: jest.fn(),
      }));

      const component = shallow(
        <Provider
          {...props}
        >
          <div>Content</div>
        </Provider>,
      );

      component.instance().componentDidMount();

      expect(component.instance().helper.search).toHaveBeenCalled();
    });
  });
});
