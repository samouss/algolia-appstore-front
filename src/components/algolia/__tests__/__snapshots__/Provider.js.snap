// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`algolia <Provider /> expect to render 1`] = `
<div>
  Content
</div>
`;
